balancer/pickfirst_leaf/pickfirst_leaf.go

```diff
@@ -0,0 +1,627 @@
+		if !b.addressIndex.increment() {
+			b.endFirstPass()
+		}
+		b.requestConnection()
+	case connectivity.Ready:
+		// Should never happen.
+		b.logger.Warningf("Requesting a connection even though we have a READY subconn")
+	}
+}
+
+func (b *pickfirstBalancer) updateSubConnState(sd *scData, state balancer.SubConnState) {
+	// Previously relevant subconns can still callback with state updates.
+	// To prevent pickers from returning these obsolete subconns, this logic
+	// is included to check if the current list of active subconns includes this
+	// subconn.
+	if activeSd, found := b.subConns.Get(*sd.addr); !found || activeSd != sd {
+		return
+	}
+	if state.ConnectivityState == connectivity.Shutdown {
+		return
+	}
+
+	if state.ConnectivityState == connectivity.Ready {
+		b.shutdownRemaining(sd)
+		b.addressIndex.seekTo(sd.addr)
+		b.state = connectivity.Ready
+		b.firstErr = nil
+		b.cc.UpdateState(balancer.State{
+			ConnectivityState: connectivity.Ready,
+			Picker:            &picker{result: balancer.PickResult{SubConn: sd.subConn}},
+		})
+		return
+	}
+
+	// If we are transitioning from READY to IDLE, shutdown and re-connect when
+	// prompted.
+	if state.ConnectivityState == connectivity.Idle && b.state == connectivity.Ready {
+		b.state = connectivity.Idle
+		b.addressIndex.reset()
+		b.cc.UpdateState(balancer.State{
+			ConnectivityState: connectivity.Idle,
+			Picker: &idlePicker{
+				exitIdle: b.ExitIdle,
+			},
+		})
+		return
+	}
+
+	if b.firstPass {
+		switch state.ConnectivityState {
+		case connectivity.Connecting:
+			if b.state == connectivity.Idle {
+				b.cc.UpdateState(balancer.State{
+					ConnectivityState: connectivity.Connecting,
+					Picker:            &picker{err: balancer.ErrNoSubConnAvailable},
+				})
+			}
+		case connectivity.TransientFailure:
+			if b.firstErr == nil {
+				b.firstErr = state.ConnectionError
+			}
+			curAddr, err := b.addressIndex.currentAddress()
+			if err != nil {
+				// This is not expected since we end the first pass when we
+				// reach the end of the list.
+				b.logger.Errorf("Current index is invalid during first pass: %v", err)
+				return
+			}
+			if activeSd, found := b.subConns.Get(*curAddr); !found || activeSd != sd {
+				return
+			}
+			if b.addressIndex.increment() {
+				b.requestConnection()
+				return
+			}
+			// End of the first pass.
+			b.endFirstPass()
+		}
+		return
+	}
+
+	// We have finished the first pass, keep re-connecting failing subconns.
+	switch state.ConnectivityState {
+	case connectivity.TransientFailure:
+		b.cc.UpdateState(balancer.State{
+			ConnectivityState: connectivity.TransientFailure,
+			Picker:            &picker{err: state.ConnectionError},
+		})
+		b.numTf++
+		// We request re-resolution when we've seen the same number of TFs as
+		// subconns. It could be that a subconn has seen multiple TFs due to
+		// differences in back-off durations, but this is a decent approximation.
+		if b.numTf >= b.subConns.Len() {
+			b.numTf = 0
+		}
```

Comment:
Where are we requesting re-resolution?

Reply:
Hmm. But then that means that every subConn will end up requesting re-resolution, right? That's not what we want. We want to request re-resolution after `num(addresses)` failures, right? I'm fine with punting this for later on, when we do other dualstack work. If we are doing that, we need to add a comment/TODO somewhere. Thanks.

Reply:
Opened https://github.com/grpc/grpc-go/issues/7534 for tracking and added a comment in clientconn.go. I'll re-introduce the re-resolution requests in the new `pick_first` once other LB policies are delegating to `pick_first` to avoid breakages.